storage/rocksdb/ha_rocksdb.cc

```diff
@@ -12179,6 +12344,156 @@ void Rdb_background_thread::run() {
   ddl_manager.persist_stats();
 }
 
+/*
+  A background thread to handle manual compactions,
+  except for dropping indexes/tables. Every second, it checks
+  pending manual compactions, and it calls CompactRange if there is.
+*/
+void Rdb_manual_compaction_thread::run() {
+  mysql_mutex_init(0, &m_mc_mutex, MY_MUTEX_INIT_FAST);
+  RDB_MUTEX_LOCK_CHECK(m_signal_mutex);
+  for (;;) {
+    if (m_stop) {
+      break;
+    }
+    timespec ts;
+    clock_gettime(CLOCK_REALTIME, &ts);
+    ts.tv_sec += 1;
+
+    const auto ret MY_ATTRIBUTE((__unused__)) =
+        mysql_cond_timedwait(&m_signal_cond, &m_signal_mutex, &ts);
+    if (m_stop) {
+      break;
+    }
+    // make sure, no program error is returned
+    DBUG_ASSERT(ret == 0 || ret == ETIMEDOUT);
+    RDB_MUTEX_UNLOCK_CHECK(m_signal_mutex);
+
+    RDB_MUTEX_LOCK_CHECK(m_mc_mutex);
+    // Grab the first item and proceed, if not empty.
+    if (m_requests.empty()) {
+      RDB_MUTEX_UNLOCK_CHECK(m_mc_mutex);
+      RDB_MUTEX_LOCK_CHECK(m_signal_mutex);
+      continue;
+    }
+    Manual_compaction_request &mcr = m_requests.begin()->second;
+    DBUG_ASSERT(mcr.cf != nullptr);
+    DBUG_ASSERT(mcr.state == Manual_compaction_request::INITED);
+    mcr.state = Manual_compaction_request::RUNNING;
+    RDB_MUTEX_UNLOCK_CHECK(m_mc_mutex);
+
+    DBUG_ASSERT(mcr.state == Manual_compaction_request::RUNNING);
+    // NO_LINT_DEBUG
+    sql_print_information("Manual Compaction id %d cf %s started.", mcr.mc_id,
+                          mcr.cf->GetName().c_str());
+    rocksdb_manual_compactions_running++;
+    if (rocksdb_debug_manual_compaction_delay > 0) {
+      // In Facebook MySQL 5.6.35, my_sleep breaks the sleep when the server
+      // gets a shutdown signal and this code depended on that behavior.
+      // In 5.7, for whatever reason, this is not the case.  my_sleep will
+      // continue to sleep until the sleep time has elapsed.  For the purpose
+      // of this variable and the accompanying test case, we need to break this
+      // down into a loop that sleeps and checks to see if the thread was
+      // signalled with the stop flag.  It is ugly, but without having DBUG_SYNC
+      // available in background threads, it is good enough for the test.
+      for (uint32_t sleeps = 0; sleeps < rocksdb_debug_manual_compaction_delay;
+           sleeps++) {
+        if (m_stop)
```

Comment:
Needs to be protected by m_signal_mutex?

Reply:
Possibly, but for this specific purpose it is only an unprotected/dirty read which shouldn't cause any real problems.  m_stop is just the indicator that the thread has been requested to terminate.  Since I am going to re-push for the other suggestions, I can go ahead and add in a lock/read/unlock around m_stop.

Reply:
I am not sure whether this is safe even accepting that it's a dirty read: what prevents compiler lifting its load out of the loop?